dot/state/block.go

```diff
@@ -431,9 +456,24 @@ func (bs *BlockState) SetFinalizedHash(hash common.Hash, round, setID uint64) er
 		bs.pruneKeyCh <- header
 	}
 
+	bs.lastFinalised = hash
```

Comment:
Can we set this before pruning and do the pruning in the background?

Reply:
the pruning needs to be synchronous, otherwise we might build on invalid chains or end up re-adding invalid chains. it doesn't make a difference whether it's before the other code or not, since this is the only function that uses `bs.lastFinalised` and it's all within a mutex